File: app-scala/app/artscentre/orm/scratch.scala
package artscentre.orm

import anorm.SqlParser._
import anorm._
import anorm.~
import scala.util.Try


object a {

  case class ProjectInfo(name: String,
                         owner: String,
                         created: java.util.Date)

  def readAllProjects(dbconn: java.sql.Connection): Try[List[ProjectInfo]] = Try {

    val query = "SELECT name, owner, created FROM projects"

    val mapping =
      get[String]("projects.owner") ~
      get[String]("projects.name") ~
      get[java.util.Date]("projects.created") map {
        case owner~name~created =>
          ProjectInfo(name, owner, created)
      }

    SQL(query).as(mapping *)(dbconn)
  }

}

object b {

  case class Skill(name: String, description: String)

  def readAllSkills(dbconn: java.sql.Connection): Try[List[Skill]] = Try {

    val query = "SELECT name, description FROM skills"

    val mapping =
      get[String]("skills.name") ~
        get[String]("skills.description") map {
        case owner~name =>
          Skill(name, owner)
      }

    SQL(query).as(mapping *)(dbconn)

  }

}


object c {

  case class Skill(name: String, description: String)

  case class ProjectInfo(name: String,
                         owner: String,
                         created: java.util.Date,
                         skill: Skill)

  def readAllProjects(dbconn: java.sql.Connection): Try[List[ProjectInfo]] = Try {

    val query =
      """
        SELECT projects.name, projects.owner, projects.created,
               skills.name, skills.description
        FROM projects
        INNER JOIN skills ON projects.skillId = skills.id
      """

    val mapping =
      get[String]("projects.name") ~
      get[String]("projects.owner") ~
      get[java.util.Date]("projects.created") ~
      get[String]("skills.name") ~
      get[String]("skills.description") map {
        case owner~name~created~skillName~skillDesc =>
          val skill = Skill(skillName, skillDesc)
          ProjectInfo(name, owner, created, skill)
      }

    SQL(query).as(mapping *)(dbconn)
  }

}

object d {

  case class Skill(name: String, description: String)

  case class ProjectInfo(name: String,
                         owner: String,
                         created: java.util.Date,
                         skills: Set[Skill])



  def readSkillsForProject(dbconn: java.sql.Connection, projectId: String): Try[Set[Skill]] = Try {

    val query =
      """
        SELECT skills.name, skills.description FROM skills
        INNER JOIN project_skills ON skills.id = project_skills.skill_id
        INNER JOIN projects ON projects.id = project_skills.project_id
        WHERE projects.id = {projectId}
      """

    val mapping =
      get[String]("skills.name") ~
      get[String]("skills.description") map {
        case name~desc =>
          Skill(name, desc)
      }

    SQL(query).on('projectId -> projectId).as(mapping *)(dbconn).toSet
  }

  def readAllProjects(dbconn: java.sql.Connection): Try[Set[ProjectInfo]] = Try {

    val query = "SELECT projcets.id, projects.name, projects.owner, projects.created FROM projects"

    val mapping =
      get[String]("projects.id") ~
      get[String]("projects.name") ~
      get[String]("projects.owner") ~
      get[java.util.Date]("projects.created") map {
        case id~name~owner~created =>
          id -> (name, owner, created)
      }


    val projects: Map[String, (String, String, java.util.Date)] = SQL(query).as(mapping *)(dbconn).toMap

    projects.map { case (projectId, project) =>
      lazy val skills: Set[Skill] = readSkillsForProject(dbconn, projectId).get
      ProjectInfo(project._1, project._2, project._3, skills)
    }.toSet
  }

}

File: app-clojure/src/artscentre/orm/scratch.clj
(ns artscentre.orm.scratch
  (:use [datomic.api :only [db q] :as d]
        [datomico.core :only [build-schema]]
        [artscentre.datomic-util]))








(def Skill (build-schema :Skill [[:name         :string :unique]
                                 [:description  :string]]))




(def ProjectInfo (build-schema :ProjectInfo [[:name     :string :unique]
                                             [:owner    :string]
                                             [:created  :instant]
                                             [:skills   :ref :many]]))





(defn read-all-projects [dbval]
  (qes '[:find ?e
         :where [?e :ProjectInfo/name]] dbval))

(defn read-skill-by-name [dbval skill-name]
  (qe '[:find ?e
         :in $ ?name
         :where [?e :Skill/name ?name]]
       dbval skill-name))




;; entry point for repl demo

(def conn (atom nil))

(defn start-dev-db []
  (reset! conn (let [uri "datomic:mem://artscentre"]
                 (d/delete-database uri)
                 (d/create-database uri)
                 (d/connect uri)))
  (let [schema-tx (concat Skill ProjectInfo)]
    @(d/transact @conn schema-tx)))


(defn load-sample-data [dbconn]

  ;; (start-dev-db)
  ;; (def dbconn @conn)
  ;; (def dbval (db dbconn))


  ;; load some skills

  (def skill-list [{:Skill/name        "Clojure"
                    :Skill/description "omg parens"}

                   {:Skill/name        "Datomic"
                    :Skill/description "i got 99 proiblems but state ain't one"}

                   {:Skill/name        "Scala"
                    :Skill/description "what is this OO malarky (just kidding)"}

                   {:Skill/name        "Java"
                    :Skill/description "every empire eventually falls"}

                   {:Skill/name        "Postgresql"
                    :Skill/description "sometimes it's cool to be old"}])

  (def txresult @(d/transact dbconn (data-with-dbid skill-list)))

  ;; (def dbval (:db-before txresult))
  (def dbval (:db-after  txresult))



  (let [clojure     (:db/id (read-skill-by-name dbval "Clojure"))
        scala       (:db/id (read-skill-by-name dbval "Scala"))
        datomic     (:db/id (read-skill-by-name dbval "Datomic"))
        java        (:db/id (read-skill-by-name dbval "Java"))
        postgresql  (:db/id (read-skill-by-name dbval "Postgresql"))

        artscentre  {:ProjectInfo/name "Artscentre"
                     :ProjectInfo/owner "Dustin"
                     :ProjectInfo/created (java.util.Date.)
                     :ProjectInfo/skills [clojure scala datomic]}

        tinyurl     {:ProjectInfo/name "Tinyurl"
                     :ProjectInfo/owner "Kevin"
                     :ProjectInfo/created (java.util.Date.)
                     :ProjectInfo/skills [java postgresql]}]

    (def txresult @(d/transact dbconn (data-with-dbid [artscentre tinyurl]a))))



  ;; (def dbval (:db-before txresult))
  (def dbval (:db-after  txresult))

  (->> (read-all-projects dbval)
     (first)
     (d/touch)
     (:ProjectInfo/skills)
     (mapv d/touch)
     )

  )





;; (read-all-projects (db @artscentre.db/conn))
